File: src/main/java/ejektaflex/kalpis/mixin/WorldRendererMixin.java
package ejektaflex.kalpis.mixin;

import ejektaflex.kalpis.event.Events;
import net.minecraft.client.render.*;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.util.math.Matrix4f;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.gen.Accessor;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

@Mixin(WorldRenderer.class)
public abstract class WorldRendererMixin {

    @Accessor
    abstract BufferBuilderStorage getBufferBuilders();

    @Inject(method = "render", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/render/VertexConsumerProvider$Immediate;draw()V", ordinal = 0))
    private void render(MatrixStack matrices, float tickDelta, long limitTime, boolean renderBlockOutline, Camera camera, GameRenderer gameRenderer, LightmapTextureManager lightmapTextureManager, Matrix4f projection, CallbackInfo ci) {
        Events.DrawScreenEvent.Companion.getDispatcher().invoker().invoke(
                new Events.DrawScreenEvent(matrices, tickDelta, camera, gameRenderer, getBufferBuilders(), projection)
        );
    }

}

File: src/main/java/ejektaflex/kalpis/mixin/WorldRendererMixin.kt
package ejektaflex.kalpis.mixin

import ejektaflex.kalpis.event.Events
import net.minecraft.client.render.*
import net.minecraft.client.util.math.Matrix4f
import net.minecraft.client.util.math.MatrixStack
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.gen.Accessor
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo


@Mixin(WorldRenderer::class)
abstract class WorldRendererMixin {

    @Accessor("bufferBuilders")
    abstract fun getBufferBuilders(): BufferBuilderStorage


    @Inject(method = ["render"],
            at = [At(value = "INVOKE", target = "Lnet/minecraft/client/render/WorldRenderer;checkEmpty(Lnet/minecraft/client/util/math/MatrixStack;)V", ordinal = 0)])
    private fun render(
            matrices: MatrixStack, tickDelta: Float, limitTime: Long, renderBlockOutline: Boolean,
            camera: Camera, gameRenderer: GameRenderer, lightmapTextureManager: LightmapTextureManager,
            projection: Matrix4f, ci: CallbackInfo
    ) {
        Events.DrawScreenEvent.Dispatcher.invoker()(Events.DrawScreenEvent(matrices, tickDelta, camera, gameRenderer, getBufferBuilders()))
    }

}